app/sdil/controllers/RegistrationController.scala

```diff
@@ -111,11 +93,20 @@ class RegistrationController(
                         } else (()).pure[WebMonad]
       isVoluntary     =  packLarge.contains(false) && useCopacker.contains(true) && (copacks, imports).isEmpty
       regDate        <- askRegDate when (!isVoluntary)
-      packSites      <- askPackSites(
-          List.empty[Site]) emptyUnless
-      (packLarge.contains(true) && packageOwn.flatten.nonEmpty) || !copacks.isEmpty
-      warehouses      <- askWarehouses emptyUnless !isVoluntary
-      contactDetails  <- ask(contactDetailsMapping, "contact-details")
+      askPackingSites = (packLarge.contains(true) && packageOwn.flatten.nonEmpty) || !copacks.isEmpty
+      extraMessages   = ExtraMessages(messages = Map("pack-at-business-address.lead" -> s"Registered address: ${fd.rosmData.address.nonEmptyLines.mkString(", ")}"))
+      useBusinessAddress <- ask(bool, "pack-at-business-address")(implicitly, implicitly, extraMessages) when askPackingSites
+        packingSites   = if (useBusinessAddress.getOrElse(false)) {
+                          List(Site.fromAddress(fd.rosmData.address))
+                         } else {
+                          List.empty[Site]
+                         }
+      firstPackingSite <- ask(packagingSiteMapping,"first-production-site")(packagingSiteForm, implicitly, ExtraMessages()) when packingSites.isEmpty
+      packSites       <- askPackSites(packingSites ++ firstPackingSite.fold(List.empty[Site])(x => List(x))) emptyUnless askPackingSites
+      addWarehouses   <- ask(bool, "ask-secondary-warehouses")(implicitly, implicitly, extraMessages) when !isVoluntary
+      firstWarehouse  <- ask(warehouseSiteMapping,"first-warehouse")(warehouseSiteForm, implicitly, ExtraMessages()) when addWarehouses.getOrElse(false)
+      warehouses      <- askWarehouses(List.empty[Site] ++ firstWarehouse.fold(List.empty[Site])(x => List(x))) emptyUnless addWarehouses.getOrElse(false)
```

Comment:
Might be worth extracting `addWarehouses`, `firstWarehhouse`, `warehouses`, `useBusinessAddress` and `packingSites` out into a separate function, perhaps?